Support patch 1.37

tfausak opened the pull request:
Fixes #47.

danielsamuels commented:
I've got another one for you:

https://media.rocketleaguereplays.com/uploads/replay_files/6B0D1614417085A7AAD82EAA30D8DABE.replay
```
>>> octane-0.20.2-osx: Data.Binary.Get.runGet at position 452067: don't know how to get attribute value Text {textValue = "TAGame.PRI_TA:BotProductName"}
CallStack (from HasCallStack):
  error, called at libraries/binary/src/Data/Binary/Get.hs:342:5 in binary-0.8.3.0:Data.Binary.Get
```

tfausak commented:
Thanks! The mighty @jjbott already has [his parser](https://github.com/jjbott/RocketLeagueReplayParser) working with 1.37. It should be pretty easy for me to, uh, borrow from him 😄

tfausak commented:
Huh, I am running into a weird problem here.

```
rattletrap:
  Data.Binary.Get.runGet at position 1077:
    unknown system id Word8 {word8Value = 161}
CallStack (from HasCallStack):
  error, called at libraries/binary/src/Data/Binary/Get.hs:342:5 in binary-0.8.3.0:Data.Binary.Get
CallStack (from -prof):
  Rattletrap.Content.getContent.frames (library/Rattletrap/Content.hs:(79,7)-(90,31))
  Rattletrap.Primitive.Int32.getInt32 (library/Rattletrap/Primitive/Int32.hs:(18,1)-(20,20))
  Rattletrap.Primitive.Text.getText (library/Rattletrap/Primitive/Text.hs:(21,1)-(27,18))
  Rattletrap.ClassMapping.getClassMapping (library/Rattletrap/ClassMapping.hs:(13,1)-(16,35))
  Rattletrap.Cache.getCache (library/Rattletrap/Cache.hs:(16,1)-(21,62))
  Rattletrap.Primitive.List.getList (library/Rattletrap/Primitive/List.hs:(13,1)-(16,22))
  Rattletrap.Content.getContent (library/Rattletrap/Content.hs:(65,1)-(103,14))
  Rattletrap.Primitive.Section.getSection.body (library/Rattletrap/Primitive/Section.hs:34:7-42)
  Rattletrap.Primitive.Word32.getWord32 (library/Rattletrap/Primitive/Word32.hs:(18,1)-(20,22))
  Rattletrap.Primitive.Section.getSection (library/Rattletrap/Primitive/Section.hs:(28,1)-(35,21))
  Rattletrap.Replay.getReplay (library/Rattletrap/Replay.hs:(23,1)-(29,30))
  Rattletrap.Helper.decodeReplay (library/Rattletrap/Helper.hs:(17,1)-(20,39))
  Rattletrap.Main.mainWithArgs (library/Rattletrap/Main.hs:(28,1)-(46,50))
  Rattletrap.Main.main (library/Rattletrap/Main.hs:(12,1)-(14,19))
```

Not sure what's going on with that.

tfausak commented:
Here's a much smaller (32 KB) replay from the latest patch: [AFB1F46A49737E36928E1EABC6F5B7AD.zip](https://github.com/tfausak/rattletrap/files/1349066/AFB1F46A49737E36928E1EABC6F5B7AD.zip)

I'm trying to figure out why my parser chokes on it. I added a bunch of logging. Something is going wrong in the very first frame. It correctly parses all the spawned replications but then has trouble with the updates. Here are the last few lines of the log:

```
    - Replication for actor 0:
        Type: Updated
        Attributes:
          Engine.GameReplicationInfo:GameClass: FlaggedIntAttributeValue (FlaggedIntAttribute {flaggedIntAttributeFlag = False, flaggedIntAttributeInt = Int32 {int32Value = 299}})
          Engine.GameReplicationInfo:ServerName: StringAttributeValue (StringAttribute {stringAttributeValue = Text {textValue = "Offline Match"}})
          ProjectX.GRI_X:bGameStarted: BooleanAttributeValue (BooleanAttribute {booleanAttributeValue = True})
          ProjectX.GRI_X:ReplicatedGamePlaylist: IntAttributeValue (IntAttribute {intAttributeValue = Int32 {int32Value = 8}})
    - Replication for actor 1:
        Type: Updated
        Attributes:
          TAGame.GameEvent_TA:ReplicatedStateName: IntAttributeValue (IntAttribute {intAttributeValue = Int32 {int32Value = 15}})
          TAGame.GameEvent_TA:BotSkill: IntAttributeValue (IntAttribute {intAttributeValue = Int32 {int32Value = 1065353216}})
          TAGame.GameEvent_TA:MatchTypeClass: FlaggedIntAttributeValue (FlaggedIntAttribute {flaggedIntAttributeFlag = False, flaggedIntAttributeInt = Int32 {int32Value = 300}})
          TAGame.GameEvent_Team_TA:MaxTeamSize: IntAttributeValue (IntAttribute {intAttributeValue = Int32 {int32Value = 1}})
          TAGame.GameEvent_Soccar_TA:SecondsRemaining: IntAttributeValue (IntAttribute {intAttributeValue = Int32 {int32Value = 300}})
    - Replication for actor 2:
        Type: Updated
        Attributes:
          TAGame.Team_TA:GameEvent: FlaggedIntAttributeValue (FlaggedIntAttribute {flaggedIntAttributeFlag = True, flaggedIntAttributeInt = Int32 {int32Value = 1}})
    - Replication for actor 3:
        Type: Updated
        Attributes:
          TAGame.Team_TA:GameEvent: FlaggedIntAttributeValue (FlaggedIntAttribute {flaggedIntAttributeFlag = True, flaggedIntAttributeInt = Int32 {int32Value = 1}})
    - Replication for actor 4:
        Type: Updated
        Attributes:
          ProjectX.GRI_X:GameServerID: QWordAttributeValue (QWordAttribute {qWordAttributeValue = Word64 {word64Value = 7594301859288842246}})
    - Replication for actor 28:
        Type: Destroyed
- Frame 1 (383 remaining):
    Time: 0.00000000000000000000000000000000000000000377
rattletrap.EXE: Data.Binary.Get.runGet at position 291: time too small
CallStack (from HasCallStack):
  error, called at libraries\binary\src\Data\Binary\Get.hs:351:5 in binary-0.8.5.1:Data.Binary.Get
```

The server ID looks bogus, so parsing might have gone off the rails before then. Everything for actor 1 looks alright. I'm still not sure what the problem is.

jjbott commented:
If it helps, the first property I see on Actor 4 is a property id 34: "Engine.PlayerReplicationInfo:PlayerName" with value "Nadir"

tfausak commented:
Thanks! Looks like parser is trying to read the wrong properties for actor 4. Do you get the same stuff for actors 0 through 3? 

Also, looking at the spawned replication for actor 4, it's a player replication info (PRI) so I don't know why it's trying to get a server ID.

```
- Frame 0 (384 remaining):
    Time: 3.8193002
    Delta: 0.0
    - Replication for actor 0:
        Type: Spawned
        Flag: False
        Name index: 0
        Name: GRI_TA_1
        Object ID: 46
        Object name: GameInfo_Soccar.GameInfo.GameInfo_Soccar:GameReplicationInfoArchetype
        Class name: TAGame.GRI_TA
        Initialization:
          Location: Vector {vectorBias = 2, vectorX = 0, vectorY = 0, vectorZ = 0}
          Rotation: n/a
    - Replication for actor 1:
        Flag: True
        Name index: 0
        Name: GRI_TA_1
        Object ID: 79
        Object name: Archetypes.GameEvent.GameEvent_Soccar
        Class name: TAGame.GameEvent_Soccar_TA
          Location: Vector {vectorBias = 2, vectorX = 0, vectorY = 0, vectorZ = 0}
          Rotation: n/a
    - Replication for actor 2:
        Flag: False
        Name index: 1
        Name: GameEvent_Soccar_TA_0
        Object ID: 91
        Object name: Archetypes.Teams.Team0
        Class name: TAGame.Team_Soccar_TA
          Location: Vector {vectorBias = 2, vectorX = 0, vectorY = 0, vectorZ = 0}
          Rotation: n/a
    - Replication for actor 3:
        Flag: True
        Name index: 1
        Name: GameEvent_Soccar_TA_0
        Object ID: 92
        Object name: Archetypes.Teams.Team1
        Class name: TAGame.Team_Soccar_TA
          Location: Vector {vectorBias = 2, vectorX = 0, vectorY = 0, vectorZ = 0}
          Rotation: n/a
    - Replication for actor 4:
        Flag: False
        Name index: 2
        Name: Team_Soccar_TA_0
        Object ID: 173
        Object name: TAGame.Default__PRI_TA
        Class name: TAGame.PRI_TA
          Location: Vector {vectorBias = 2, vectorX = 0, vectorY = 0, vectorZ = 0}
          Rotation: n/a
```

This might be a problem with the class attribute map, which would suck. Debugging that thing is a pain in the butt.

jjbott commented:
Yeah, 0 through 3 look fine. I'm betting it is a problem with the attribute map. Both ProjectX.GRI_X:GameServerID and Engine.PlayerReplicationInfo:PlayerName are id 34. My class parent fixinator is getting triggered on TAame.PRI_TA:

> Fixing class ProjectX.PRI_X, setting its parent to Engine.PlayerReplicationInfo from NULL
> Fixing class TAGame.PRI_TA, setting its parent to ProjectX.PRI_X from NULL

The whole mess looks like this:

```
ClassNetCache: ObjectIndex 0 (Core.Object ParentId 0 Id 0
    ClassNetCache: ObjectIndex 22 (Engine.Actor ParentId 0 Id 21
        ClassNetCacheProperty: Index 1 (Engine.Actor:RelativeRotation) Id 0
        ClassNetCacheProperty: Index 2 (Engine.Actor:RelativeLocation) Id 1
        ClassNetCacheProperty: Index 3 (Engine.Actor:Velocity) Id 2
        ClassNetCacheProperty: Index 4 (Engine.Actor:Instigator) Id 3
        ClassNetCacheProperty: Index 5 (Engine.Actor:bNetOwner) Id 4
        ClassNetCacheProperty: Index 6 (Engine.Actor:bProjTarget) Id 5
        ClassNetCacheProperty: Index 7 (Engine.Actor:bBlockActors) Id 6
        ClassNetCacheProperty: Index 8 (Engine.Actor:bCollideWorld) Id 7
        ClassNetCacheProperty: Index 9 (Engine.Actor:bCollideActors) Id 8
        ClassNetCacheProperty: Index 10 (Engine.Actor:bHardAttach) Id 9
        ClassNetCacheProperty: Index 11 (Engine.Actor:bTearOff) Id 10
        ClassNetCacheProperty: Index 12 (Engine.Actor:bHidden) Id 11
        ClassNetCacheProperty: Index 13 (Engine.Actor:Base) Id 12
        ClassNetCacheProperty: Index 14 (Engine.Actor:Owner) Id 13
        ClassNetCacheProperty: Index 15 (Engine.Actor:ReplicatedCollisionType) Id 14
        ClassNetCacheProperty: Index 16 (Engine.Actor:Role) Id 15
        ClassNetCacheProperty: Index 17 (Engine.Actor:RemoteRole) Id 16
        ClassNetCacheProperty: Index 18 (Engine.Actor:Physics) Id 17
        ClassNetCacheProperty: Index 19 (Engine.Actor:DrawScale) Id 18
        ClassNetCacheProperty: Index 20 (Engine.Actor:Rotation) Id 19
        ClassNetCacheProperty: Index 21 (Engine.Actor:Location) Id 20
        ClassNetCache: ObjectIndex 23 (Engine.Info ParentId 21 Id 21
            ClassNetCache: ObjectIndex 24 (Engine.ReplicationInfo ParentId 21 Id 21
                ClassNetCache: ObjectIndex 36 (Engine.GameReplicationInfo ParentId 21 Id 32
                    ClassNetCacheProperty: Index 25 (Engine.GameReplicationInfo:Winner) Id 21
                    ClassNetCacheProperty: Index 26 (Engine.GameReplicationInfo:ServerName) Id 22
                    ClassNetCacheProperty: Index 27 (Engine.GameReplicationInfo:TimeLimit) Id 23
                    ClassNetCacheProperty: Index 28 (Engine.GameReplicationInfo:GoalScore) Id 24
                    ClassNetCacheProperty: Index 29 (Engine.GameReplicationInfo:RemainingMinute) Id 25
                    ClassNetCacheProperty: Index 30 (Engine.GameReplicationInfo:ElapsedTime) Id 26
                    ClassNetCacheProperty: Index 31 (Engine.GameReplicationInfo:RemainingTime) Id 27
                    ClassNetCacheProperty: Index 32 (Engine.GameReplicationInfo:bMatchIsOver) Id 28
                    ClassNetCacheProperty: Index 33 (Engine.GameReplicationInfo:bMatchHasBegun) Id 29
                    ClassNetCacheProperty: Index 34 (Engine.GameReplicationInfo:bStopCountDown) Id 30
                    ClassNetCacheProperty: Index 35 (Engine.GameReplicationInfo:GameClass) Id 31
                    ClassNetCache: ObjectIndex 43 (ProjectX.GRI_X ParentId 32 Id 38
                        ClassNetCacheProperty: Index 37 (ProjectX.GRI_X:MatchGUID) Id 32
                        ClassNetCacheProperty: Index 38 (ProjectX.GRI_X:bGameStarted) Id 33
                        ClassNetCacheProperty: Index 39 (ProjectX.GRI_X:GameServerID) Id 34
                        ClassNetCacheProperty: Index 40 (ProjectX.GRI_X:Reservations) Id 35
                        ClassNetCacheProperty: Index 41 (ProjectX.GRI_X:ReplicatedGameMutatorIndex) Id 36
                        ClassNetCacheProperty: Index 42 (ProjectX.GRI_X:ReplicatedGamePlaylist) Id 37
                        ClassNetCache: ObjectIndex 45 (TAGame.GRI_TA ParentId 38 Id 39
                            ClassNetCacheProperty: Index 44 (TAGame.GRI_TA:NewDedicatedServerIP) Id 38
                ClassNetCache: ObjectIndex 83 (Engine.TeamInfo ParentId 21 Id 24
                    ClassNetCacheProperty: Index 80 (Engine.TeamInfo:TeamIndex) Id 21
                    ClassNetCacheProperty: Index 81 (Engine.TeamInfo:Score) Id 22
                    ClassNetCacheProperty: Index 82 (Engine.TeamInfo:TeamName) Id 23
                    ClassNetCache: ObjectIndex 88 (TAGame.Team_TA ParentId 24 Id 28
                        ClassNetCacheProperty: Index 84 (TAGame.Team_TA:ClubColors) Id 24
                        ClassNetCacheProperty: Index 85 (TAGame.Team_TA:LogoData) Id 25
                        ClassNetCacheProperty: Index 86 (TAGame.Team_TA:CustomTeamName) Id 26
                        ClassNetCacheProperty: Index 87 (TAGame.Team_TA:GameEvent) Id 27
                        ClassNetCache: ObjectIndex 90 (TAGame.Team_Soccar_TA ParentId 28 Id 29
                            ClassNetCacheProperty: Index 89 (TAGame.Team_Soccar_TA:GameScore) Id 28
                ClassNetCache: ObjectIndex 112 (Engine.PlayerReplicationInfo ParentId 21 Id 38
                    ClassNetCacheProperty: Index 93 (Engine.PlayerReplicationInfo:RemoteUserData) Id 21
                    ClassNetCacheProperty: Index 94 (Engine.PlayerReplicationInfo:UniqueId) Id 22
                    ClassNetCacheProperty: Index 95 (Engine.PlayerReplicationInfo:bFromPreviousLevel) Id 23
                    ClassNetCacheProperty: Index 96 (Engine.PlayerReplicationInfo:bIsInactive) Id 24
                    ClassNetCacheProperty: Index 97 (Engine.PlayerReplicationInfo:bBot) Id 25
                    ClassNetCacheProperty: Index 98 (Engine.PlayerReplicationInfo:bOutOfLives) Id 26
                    ClassNetCacheProperty: Index 99 (Engine.PlayerReplicationInfo:bReadyToPlay) Id 27
                    ClassNetCacheProperty: Index 100 (Engine.PlayerReplicationInfo:bWaitingPlayer) Id 28
                    ClassNetCacheProperty: Index 101 (Engine.PlayerReplicationInfo:bOnlySpectator) Id 29
                    ClassNetCacheProperty: Index 102 (Engine.PlayerReplicationInfo:bIsSpectator) Id 30
                    ClassNetCacheProperty: Index 103 (Engine.PlayerReplicationInfo:bAdmin) Id 31
                    ClassNetCacheProperty: Index 104 (Engine.PlayerReplicationInfo:Team) Id 32
                    ClassNetCacheProperty: Index 105 (Engine.PlayerReplicationInfo:PlayerID) Id 33
                    ClassNetCacheProperty: Index 106 (Engine.PlayerReplicationInfo:PlayerName) Id 34
                    ClassNetCacheProperty: Index 107 (Engine.PlayerReplicationInfo:Ping) Id 35
                    ClassNetCacheProperty: Index 108 (Engine.PlayerReplicationInfo:Deaths) Id 36
                    ClassNetCacheProperty: Index 109 (Engine.PlayerReplicationInfo:Score) Id 37
                    ClassNetCacheProperty: Index 110 (Engine.PlayerReplicationInfo:ClientFillRemoteUserData) Id 38
                    ClassNetCacheProperty: Index 111 (Engine.PlayerReplicationInfo:ServerUpdateRemoteUserData) Id 39
                    ClassNetCache: ObjectIndex 113 (ProjectX.PRI_X ParentId 40 Id 38
                        ClassNetCache: ObjectIndex 172 (TAGame.PRI_TA ParentId 40 Id 73
                            ClassNetCacheProperty: Index 114 (TAGame.PRI_TA:BotProductName) Id 40
                            ClassNetCacheProperty: Index 115 (TAGame.PRI_TA:RepStatTitles) Id 41
                            ClassNetCacheProperty: Index 116 (TAGame.PRI_TA:ReplacingBotPRI) Id 42
                            ClassNetCacheProperty: Index 117 (TAGame.PRI_TA:MaxTimeTillItem) Id 43
                            ClassNetCacheProperty: Index 118 (TAGame.PRI_TA:TimeTillItem) Id 44
                            ClassNetCacheProperty: Index 119 (TAGame.PRI_TA:SteeringSensitivity) Id 45
                            ClassNetCacheProperty: Index 120 (TAGame.PRI_TA:SkillTier) Id 46
                            ClassNetCacheProperty: Index 121 (TAGame.PRI_TA:Title) Id 47
                            ClassNetCacheProperty: Index 122 (TAGame.PRI_TA:TotalXP) Id 48
                            ClassNetCacheProperty: Index 123 (TAGame.PRI_TA:PartyLeader) Id 49
                            ClassNetCacheProperty: Index 124 (TAGame.PRI_TA:PlayerHistoryKey) Id 50
                            ClassNetCacheProperty: Index 125 (TAGame.PRI_TA:PawnType) Id 51
                            ClassNetCacheProperty: Index 126 (TAGame.PRI_TA:PersistentCamera) Id 52
                            ClassNetCacheProperty: Index 127 (TAGame.PRI_TA:RespawnTimeRemaining) Id 53
                            ClassNetCacheProperty: Index 128 (TAGame.PRI_TA:ClientLoadoutsOnline) Id 54
                            ClassNetCacheProperty: Index 129 (TAGame.PRI_TA:ClientLoadouts) Id 55
                            ClassNetCacheProperty: Index 130 (TAGame.PRI_TA:ClientLoadoutOnline) Id 56
                            ClassNetCacheProperty: Index 131 (TAGame.PRI_TA:ClientLoadout) Id 57
                            ClassNetCacheProperty: Index 132 (TAGame.PRI_TA:ReplicatedGameEvent) Id 58
                            ClassNetCacheProperty: Index 133 (TAGame.PRI_TA:PlayerHistoryValid) Id 59
                            ClassNetCacheProperty: Index 134 (TAGame.PRI_TA:bUsingItems) Id 60
                            ClassNetCacheProperty: Index 135 (TAGame.PRI_TA:bVoteToForfeitDisabled) Id 61
                            ClassNetCacheProperty: Index 136 (TAGame.PRI_TA:bIsInSplitScreen) Id 62
                            ClassNetCacheProperty: Index 137 (TAGame.PRI_TA:bBusy) Id 63
                            ClassNetCacheProperty: Index 138 (TAGame.PRI_TA:bReady) Id 64
                            ClassNetCacheProperty: Index 139 (TAGame.PRI_TA:bOnlineLoadoutsSet) Id 65
                            ClassNetCacheProperty: Index 140 (TAGame.PRI_TA:bOnlineLoadoutSet) Id 66
                            ClassNetCacheProperty: Index 141 (TAGame.PRI_TA:bMatchAdmin) Id 67
                            ClassNetCacheProperty: Index 142 (TAGame.PRI_TA:bMatchMVP) Id 68
                            ClassNetCacheProperty: Index 143 (TAGame.PRI_TA:MatchBreakoutDamage) Id 69
                            ClassNetCacheProperty: Index 144 (TAGame.PRI_TA:MatchShots) Id 70
                            ClassNetCacheProperty: Index 145 (TAGame.PRI_TA:MatchSaves) Id 71
                            ClassNetCacheProperty: Index 146 (TAGame.PRI_TA:MatchAssists) Id 72
                            ClassNetCacheProperty: Index 147 (TAGame.PRI_TA:MatchGoals) Id 73
                            ClassNetCacheProperty: Index 148 (TAGame.PRI_TA:MatchScore) Id 74
                            ClassNetCacheProperty: Index 149 (TAGame.PRI_TA:ServerSetPlayerHistoryKey) Id 75
                            ClassNetCacheProperty: Index 150 (TAGame.PRI_TA:ServerSetPartyLeader) Id 76
                            ClassNetCacheProperty: Index 151 (TAGame.PRI_TA:ServerSplitScreenStatusChanged) Id 77
                            ClassNetCacheProperty: Index 152 (TAGame.PRI_TA:ServerSetLoadout) Id 78
                            ClassNetCacheProperty: Index 153 (TAGame.PRI_TA:ClientNotifyGainedStat) Id 79
                            ClassNetCacheProperty: Index 154 (TAGame.PRI_TA:ClientScoredGoal) Id 80
                            ClassNetCacheProperty: Index 155 (TAGame.PRI_TA:ClientNotifyStatTickerMessage) Id 81
                            ClassNetCacheProperty: Index 156 (TAGame.PRI_TA:ClientGiveOnlineReward) Id 82
                            ClassNetCacheProperty: Index 157 (TAGame.PRI_TA:ClientGenerateItemDrop) Id 83
                            ClassNetCacheProperty: Index 158 (TAGame.PRI_TA:ClientGiveOnlineCurrencyDrop) Id 84
                            ClassNetCacheProperty: Index 159 (TAGame.PRI_TA:ClientUpdateOnlineProductStats) Id 85
                            ClassNetCacheProperty: Index 160 (TAGame.PRI_TA:ServerReadyUp) Id 86
                            ClassNetCacheProperty: Index 161 (TAGame.PRI_TA:ServerMatchAdminSetScoreAndTime) Id 87
                            ClassNetCacheProperty: Index 162 (TAGame.PRI_TA:ServerMatchAdminSetPaused) Id 88
                            ClassNetCacheProperty: Index 163 (TAGame.PRI_TA:ServerSetVoteStatus) Id 89
                            ClassNetCacheProperty: Index 164 (TAGame.PRI_TA:ServerChangeTeam) Id 90
                            ClassNetCacheProperty: Index 165 (TAGame.PRI_TA:ClientChangeTeamFailed) Id 91
                            ClassNetCacheProperty: Index 166 (TAGame.PRI_TA:ServerSpectate) Id 92
                            ClassNetCacheProperty: Index 167 (TAGame.PRI_TA:ServerSetTotalXP) Id 93
                            ClassNetCacheProperty: Index 168 (TAGame.PRI_TA:ServerSetUserCarPreferences) Id 94
                            ClassNetCacheProperty: Index 169 (TAGame.PRI_TA:ServerVoteToForfeit) Id 95
                            ClassNetCacheProperty: Index 170 (TAGame.PRI_TA:ServerSetPlayerActivatedFX) Id 96
                            ClassNetCacheProperty: Index 171 (TAGame.PRI_TA:ClientUnlockAchievement) Id 97
                ClassNetCache: ObjectIndex 187 (TAGame.CameraSettingsActor_TA ParentId 21 Id 28
                    ClassNetCacheProperty: Index 174 (TAGame.CameraSettingsActor_TA:CameraYaw) Id 21
                    ClassNetCacheProperty: Index 175 (TAGame.CameraSettingsActor_TA:CameraPitch) Id 22
                    ClassNetCacheProperty: Index 176 (TAGame.CameraSettingsActor_TA:bUsingFreecam) Id 23
                    ClassNetCacheProperty: Index 177 (TAGame.CameraSettingsActor_TA:bUsingBehindView) Id 24
                    ClassNetCacheProperty: Index 178 (TAGame.CameraSettingsActor_TA:bUsingSecondaryCamera) Id 25
                    ClassNetCacheProperty: Index 179 (TAGame.CameraSettingsActor_TA:ProfileSettings) Id 26
                    ClassNetCacheProperty: Index 180 (TAGame.CameraSettingsActor_TA:PRI) Id 27
                    ClassNetCacheProperty: Index 181 (TAGame.CameraSettingsActor_TA:ServerSetCameraSettings) Id 28
                    ClassNetCacheProperty: Index 182 (TAGame.CameraSettingsActor_TA:ServerSetUsingSecondaryCamera) Id 29
                    ClassNetCacheProperty: Index 183 (TAGame.CameraSettingsActor_TA:ServerSetUsingBehindView) Id 30
                    ClassNetCacheProperty: Index 184 (TAGame.CameraSettingsActor_TA:ServerSetUsingFreecam) Id 31
                    ClassNetCacheProperty: Index 185 (TAGame.CameraSettingsActor_TA:ServerSetCameraRotation) Id 32
                    ClassNetCacheProperty: Index 186 (TAGame.CameraSettingsActor_TA:ServerSetCameraRotationReliable) Id 33
                ClassNetCache: ObjectIndex 218 (Engine.Pawn ParentId 21 Id 50
                    ClassNetCacheProperty: Index 189 (Engine.Pawn:RootMotionInterpCurveLastValue) Id 21
                    ClassNetCacheProperty: Index 190 (Engine.Pawn:RootMotionInterpCurrentTime) Id 22
                    ClassNetCacheProperty: Index 191 (Engine.Pawn:RootMotionInterpRate) Id 23
                    ClassNetCacheProperty: Index 192 (Engine.Pawn:FlashLocation) Id 24
                    ClassNetCacheProperty: Index 193 (Engine.Pawn:InvManager) Id 25
                    ClassNetCacheProperty: Index 194 (Engine.Pawn:DrivenVehicle) Id 26
                    ClassNetCacheProperty: Index 195 (Engine.Pawn:TearOffMomentum) Id 27
                    ClassNetCacheProperty: Index 196 (Engine.Pawn:HitDamageType) Id 28
                    ClassNetCacheProperty: Index 197 (Engine.Pawn:TakeHitLocation) Id 29
                    ClassNetCacheProperty: Index 198 (Engine.Pawn:PlayerReplicationInfo) Id 30
                    ClassNetCacheProperty: Index 199 (Engine.Pawn:HealthMax) Id 31
                    ClassNetCacheProperty: Index 200 (Engine.Pawn:Health) Id 32
                    ClassNetCacheProperty: Index 201 (Engine.Pawn:AirControl) Id 33
                    ClassNetCacheProperty: Index 202 (Engine.Pawn:JumpZ) Id 34
                    ClassNetCacheProperty: Index 203 (Engine.Pawn:AccelRate) Id 35
                    ClassNetCacheProperty: Index 204 (Engine.Pawn:AirSpeed) Id 36
                    ClassNetCacheProperty: Index 205 (Engine.Pawn:WaterSpeed) Id 37
                    ClassNetCacheProperty: Index 206 (Engine.Pawn:GroundSpeed) Id 38
                    ClassNetCacheProperty: Index 207 (Engine.Pawn:FiringMode) Id 39
                    ClassNetCacheProperty: Index 208 (Engine.Pawn:FlashCount) Id 40
                    ClassNetCacheProperty: Index 209 (Engine.Pawn:RemoteViewPitch) Id 41
                    ClassNetCacheProperty: Index 210 (Engine.Pawn:bFastAttachedMove) Id 42
                    ClassNetCacheProperty: Index 211 (Engine.Pawn:bRootMotionFromInterpCurve) Id 43
                    ClassNetCacheProperty: Index 212 (Engine.Pawn:bUsedByMatinee) Id 44
                    ClassNetCacheProperty: Index 213 (Engine.Pawn:bCanSwatTurn) Id 45
                    ClassNetCacheProperty: Index 214 (Engine.Pawn:bSimulateGravity) Id 46
                    ClassNetCacheProperty: Index 215 (Engine.Pawn:bIsCrouched) Id 47
                    ClassNetCacheProperty: Index 216 (Engine.Pawn:bIsWalking) Id 48
                    ClassNetCacheProperty: Index 217 (Engine.Pawn:Controller) Id 49
                    ClassNetCache: ObjectIndex 219 (ProjectX.Pawn_X ParentId 50 Id 50
                        ClassNetCache: ObjectIndex 225 (TAGame.RBActor_TA ParentId 50 Id 55
                            ClassNetCacheProperty: Index 220 (TAGame.RBActor_TA:WeldedInfo) Id 50
                            ClassNetCacheProperty: Index 221 (TAGame.RBActor_TA:ReplicatedRBState) Id 51
                            ClassNetCacheProperty: Index 222 (TAGame.RBActor_TA:bIgnoreSyncing) Id 52
                            ClassNetCacheProperty: Index 223 (TAGame.RBActor_TA:bFrozen) Id 53
                            ClassNetCacheProperty: Index 224 (TAGame.RBActor_TA:bReplayActor) Id 54
                            ClassNetCache: ObjectIndex 239 (TAGame.Ball_TA ParentId 55 Id 68
                                ClassNetCacheProperty: Index 226 (TAGame.Ball_TA:BallTouchScore) Id 55
                                ClassNetCacheProperty: Index 227 (TAGame.Ball_TA:BallBelongsTo) Id 56
                                ClassNetCacheProperty: Index 228 (TAGame.Ball_TA:ReplicatedExplosionDataExtended) Id 57
                                ClassNetCacheProperty: Index 229 (TAGame.Ball_TA:ReplicatedExplosionData) Id 58
                                ClassNetCacheProperty: Index 230 (TAGame.Ball_TA:GameEvent) Id 59
                                ClassNetCacheProperty: Index 231 (TAGame.Ball_TA:HitTeamNum) Id 60
                                ClassNetCacheProperty: Index 232 (TAGame.Ball_TA:ReplicatedAddedCarBounceScale) Id 61
                                ClassNetCacheProperty: Index 233 (TAGame.Ball_TA:ReplicatedBallMaxLinearSpeedScale) Id 62
                                ClassNetCacheProperty: Index 234 (TAGame.Ball_TA:ReplicatedBallGravityScale) Id 63
                                ClassNetCacheProperty: Index 235 (TAGame.Ball_TA:ReplicatedWorldBounceScale) Id 64
                                ClassNetCacheProperty: Index 236 (TAGame.Ball_TA:ReplicatedBallMesh) Id 65
                                ClassNetCacheProperty: Index 237 (TAGame.Ball_TA:ReplicatedBallScale) Id 66
                                ClassNetCacheProperty: Index 238 (TAGame.Ball_TA:bEndOfGameHidden) Id 67
                            ClassNetCache: ObjectIndex 246 (TAGame.Vehicle_TA ParentId 55 Id 60
                                ClassNetCacheProperty: Index 241 (TAGame.Vehicle_TA:ReplicatedSteer) Id 55
                                ClassNetCacheProperty: Index 242 (TAGame.Vehicle_TA:ReplicatedThrottle) Id 56
                                ClassNetCacheProperty: Index 243 (TAGame.Vehicle_TA:bPodiumMode) Id 57
                                ClassNetCacheProperty: Index 244 (TAGame.Vehicle_TA:bReplicatedHandbrake) Id 58
                                ClassNetCacheProperty: Index 245 (TAGame.Vehicle_TA:bDriving) Id 59
                                ClassNetCache: ObjectIndex 258 (TAGame.Car_TA ParentId 60 Id 70
                                    ClassNetCacheProperty: Index 247 (TAGame.Car_TA:AddedCarForceMultiplier) Id 60
                                    ClassNetCacheProperty: Index 248 (TAGame.Car_TA:AddedBallForceMultiplier) Id 61
                                    ClassNetCacheProperty: Index 249 (TAGame.Car_TA:AttachedPickup) Id 62
                                    ClassNetCacheProperty: Index 250 (TAGame.Car_TA:AttachedBall) Id 63
                                    ClassNetCacheProperty: Index 251 (TAGame.Car_TA:ReplicatedDemolish) Id 64
                                    ClassNetCacheProperty: Index 252 (TAGame.Car_TA:bOverrideBoostOn) Id 65
                                    ClassNetCacheProperty: Index 253 (TAGame.Car_TA:bOverrideHandbrakeOn) Id 66
                                    ClassNetCacheProperty: Index 254 (TAGame.Car_TA:ReplicatedCarScale) Id 67
                                    ClassNetCacheProperty: Index 255 (TAGame.Car_TA:ClubColors) Id 68
                                    ClassNetCacheProperty: Index 256 (TAGame.Car_TA:TeamPaint) Id 69
                                    ClassNetCacheProperty: Index 257 (TAGame.Car_TA:ServerEquipProduct) Id 70
                ClassNetCache: ObjectIndex 263 (TAGame.CarComponent_TA ParentId 21 Id 24
                    ClassNetCacheProperty: Index 260 (TAGame.CarComponent_TA:ReplicatedActivityTime) Id 21
                    ClassNetCacheProperty: Index 261 (TAGame.CarComponent_TA:Vehicle) Id 22
                    ClassNetCacheProperty: Index 262 (TAGame.CarComponent_TA:ReplicatedActive) Id 23
                    ClassNetCache: ObjectIndex 275 (TAGame.CarComponent_Boost_TA ParentId 24 Id 32
                        ClassNetCacheProperty: Index 264 (TAGame.CarComponent_Boost_TA:ReplicatedBoostAmount) Id 24
                        ClassNetCacheProperty: Index 265 (TAGame.CarComponent_Boost_TA:bNoBoost) Id 25
                        ClassNetCacheProperty: Index 266 (TAGame.CarComponent_Boost_TA:UnlimitedBoostRefCount) Id 26
                        ClassNetCacheProperty: Index 267 (TAGame.CarComponent_Boost_TA:RechargeDelay) Id 27
                        ClassNetCacheProperty: Index 268 (TAGame.CarComponent_Boost_TA:RechargeRate) Id 28
                        ClassNetCacheProperty: Index 269 (TAGame.CarComponent_Boost_TA:BoostModifier) Id 29
                        ClassNetCacheProperty: Index 270 (TAGame.CarComponent_Boost_TA:CurrentBoostAmount) Id 30
                        ClassNetCacheProperty: Index 271 (TAGame.CarComponent_Boost_TA:StartBoostAmount) Id 31
                        ClassNetCacheProperty: Index 272 (TAGame.CarComponent_Boost_TA:ServerConfirmBoostAmount) Id 32
                        ClassNetCacheProperty: Index 273 (TAGame.CarComponent_Boost_TA:ClientFixBoostAmount) Id 33
                        ClassNetCacheProperty: Index 274 (TAGame.CarComponent_Boost_TA:ClientGiveBoost) Id 34
                    ClassNetCache: ObjectIndex 277 (TAGame.CarComponent_Jump_TA ParentId 24 Id 24
                    ClassNetCache: ObjectIndex 286 (TAGame.CarComponent_FlipCar_TA ParentId 24 Id 26
                        ClassNetCacheProperty: Index 284 (TAGame.CarComponent_FlipCar_TA:bFlipRight) Id 24
                        ClassNetCacheProperty: Index 285 (TAGame.CarComponent_FlipCar_TA:FlipCarTime) Id 25
                    ClassNetCache: ObjectIndex 282 (TAGame.CarComponent_Dodge_TA ParentId 24 Id 25
                        ClassNetCacheProperty: Index 281 (TAGame.CarComponent_Dodge_TA:DodgeTorque) Id 24
                    ClassNetCache: ObjectIndex 279 (TAGame.CarComponent_DoubleJump_TA ParentId 24 Id 24
                ClassNetCache: ObjectIndex 293 (TAGame.CrowdActor_TA ParentId 21 Id 26
                    ClassNetCacheProperty: Index 288 (TAGame.CrowdActor_TA:ReplicatedRoundCountDownNumber) Id 21
                    ClassNetCacheProperty: Index 289 (TAGame.CrowdActor_TA:ReplicatedCountDownNumber) Id 22
                    ClassNetCacheProperty: Index 290 (TAGame.CrowdActor_TA:ReplicatedOneShotSound) Id 23
                    ClassNetCacheProperty: Index 291 (TAGame.CrowdActor_TA:ModifiedNoise) Id 24
                    ClassNetCacheProperty: Index 292 (TAGame.CrowdActor_TA:GameEvent) Id 25
                ClassNetCache: ObjectIndex 297 (TAGame.CrowdManager_TA ParentId 21 Id 23
                    ClassNetCacheProperty: Index 295 (TAGame.CrowdManager_TA:GameEvent) Id 21
                    ClassNetCacheProperty: Index 296 (TAGame.CrowdManager_TA:ReplicatedGlobalOneShotSound) Id 22
                ClassNetCache: ObjectIndex 305 (TAGame.VehiclePickup_TA ParentId 21 Id 23
                    ClassNetCacheProperty: Index 303 (TAGame.VehiclePickup_TA:bNoPickup) Id 21
                    ClassNetCacheProperty: Index 304 (TAGame.VehiclePickup_TA:ReplicatedPickupData) Id 22
                    ClassNetCache: ObjectIndex 306 (TAGame.VehiclePickup_Boost_TA ParentId 23 Id 23
        ClassNetCache: ObjectIndex 58 (TAGame.GameEvent_TA ParentId 21 Id 32
            ClassNetCacheProperty: Index 47 (TAGame.GameEvent_TA:MatchSettings) Id 21
            ClassNetCacheProperty: Index 48 (TAGame.GameEvent_TA:GameOwner) Id 22
            ClassNetCacheProperty: Index 49 (TAGame.GameEvent_TA:ReplicatedGameStateTimeRemaining) Id 23
            ClassNetCacheProperty: Index 50 (TAGame.GameEvent_TA:ReplicatedStateName) Id 24
            ClassNetCacheProperty: Index 51 (TAGame.GameEvent_TA:ActivatorCar) Id 25
            ClassNetCacheProperty: Index 52 (TAGame.GameEvent_TA:BotSkill) Id 26
            ClassNetCacheProperty: Index 53 (TAGame.GameEvent_TA:bCanVoteToForfeit) Id 27
            ClassNetCacheProperty: Index 54 (TAGame.GameEvent_TA:bHasLeaveMatchPenalty) Id 28
            ClassNetCacheProperty: Index 55 (TAGame.GameEvent_TA:bAllowReadyUp) Id 29
            ClassNetCacheProperty: Index 56 (TAGame.GameEvent_TA:ReplicatedStateIndex) Id 30
            ClassNetCacheProperty: Index 57 (TAGame.GameEvent_TA:MatchTypeClass) Id 31
            ClassNetCache: ObjectIndex 62 (TAGame.GameEvent_Team_TA ParentId 32 Id 35
                ClassNetCacheProperty: Index 59 (TAGame.GameEvent_Team_TA:bForfeit) Id 32
                ClassNetCacheProperty: Index 60 (TAGame.GameEvent_Team_TA:bDisableMutingOtherTeam) Id 33
                ClassNetCacheProperty: Index 61 (TAGame.GameEvent_Team_TA:MaxTeamSize) Id 34
                ClassNetCache: ObjectIndex 78 (TAGame.GameEvent_Soccar_TA ParentId 35 Id 50
                    ClassNetCacheProperty: Index 63 (TAGame.GameEvent_Soccar_TA:SubRulesArchetype) Id 35
                    ClassNetCacheProperty: Index 64 (TAGame.GameEvent_Soccar_TA:ReplicatedMusicStinger) Id 36
                    ClassNetCacheProperty: Index 65 (TAGame.GameEvent_Soccar_TA:RoundNum) Id 37
                    ClassNetCacheProperty: Index 66 (TAGame.GameEvent_Soccar_TA:ReplicatedScoredOnTeam) Id 38
                    ClassNetCacheProperty: Index 67 (TAGame.GameEvent_Soccar_TA:MVP) Id 39
                    ClassNetCacheProperty: Index 68 (TAGame.GameEvent_Soccar_TA:MatchWinner) Id 40
                    ClassNetCacheProperty: Index 69 (TAGame.GameEvent_Soccar_TA:GameWinner) Id 41
                    ClassNetCacheProperty: Index 70 (TAGame.GameEvent_Soccar_TA:ReplayDirector) Id 42
                    ClassNetCacheProperty: Index 71 (TAGame.GameEvent_Soccar_TA:bMatchEnded) Id 43
                    ClassNetCacheProperty: Index 72 (TAGame.GameEvent_Soccar_TA:bNoContest) Id 44
                    ClassNetCacheProperty: Index 73 (TAGame.GameEvent_Soccar_TA:bOverTime) Id 45
                    ClassNetCacheProperty: Index 74 (TAGame.GameEvent_Soccar_TA:bBallHasBeenHit) Id 46
                    ClassNetCacheProperty: Index 75 (TAGame.GameEvent_Soccar_TA:SecondsRemaining) Id 47
                    ClassNetCacheProperty: Index 76 (TAGame.GameEvent_Soccar_TA:GameTime) Id 48
                    ClassNetCacheProperty: Index 77 (TAGame.GameEvent_Soccar_TA:SeriesLength) Id 49
```

tfausak commented:
Oh boy. I thought I'd never have to touch the class attribute map again. I'll dig into this later today.

tfausak commented:
Alright, I did a bit of digging. My problem is somewhere in `getParentClass`: https://github.com/tfausak/rattletrap/blob/a6f127f6de018b8919b2395faa93a5a6c1efe6b2/library/Rattletrap/ClassAttributeMap.hs#L177

That tries to look up the parent by name, fails for some reason, then falls back to scanning by ID. That returns the wrong parent (`TAGame.GRI_TA`, class ID 45).

```
getting parent class for TAGame.PRI_TA with parent cache id 40
getting parent class by id for parent cache id 40
getting parent class by id for parent cache id 39
found parent class id 45 for parent cache id 39
getting parent class for ProjectX.PRI_X with parent cache id 40
getting parent class by id for parent cache id 40
getting parent class by id for parent cache id 39
found parent class id 45 for parent cache id 39
```

So I guess the problem is with `getParentClassByName`. All the names are wired up correctly, but it's failing to use the information in any reasonable way.

tfausak commented on library/Rattletrap/ClassAttributeMap.hs:
```diff
@@ -205,7 +201,7 @@ getParentClassByName className parentCacheId xs =
               (\(_, parentClassId, _, _) -> parentClassId)
               (filter
-                 (\(_, _, cacheId, _) -> cacheId == parentCacheId)
+                 (\(_, _, cacheId, _) -> cacheId <= parentCacheId)
```
😆 This worked great for thousands of replays, then all of a sudden it broke. At least I only had to change one character to fix it!

tfausak commented on library/Rattletrap/Attribute/LoadoutOnline.hs:
```diff
@@ -5,37 +5,107 @@ import Rattletrap.Primitive
+  { loadoutAttributeValue :: [[ProductAttribute]]
+  } deriving (Eq, Ord, Show)
+
+data ProductAttribute = ProductAttribute
```
This should really go into its own module.
